plugins/catalog-backend-module-incremental-ingestion/src/types.ts

```diff
@@ -75,6 +75,19 @@ export interface IncrementalEntityProvider<TCursor, TContext> {
    * @param burst - a function which performs a series of iterations
    */
   around(burst: (context: TContext) => Promise<void>): Promise<void>;
+
+  /**
+   * If present, this method maps incoming payloads to apply updates
+   * outside of the incremental ingestion schedule.
+   */
+  deltaMapper?: (payload: TInput) => {
```

Comment:
I think this might be a bit too limiting.

As mentioned above, I think it'll be common that you want to listen to multiple topics (with different payloads). And the subscribers [actually receive](https://github.com/backstage/backstage/blob/d16af4a53aaa52210f8e444bb314a377aa0d868d/plugins/events-node/api-report.md?plain=1#L63) an event that has both [payload and headers](https://github.com/backstage/backstage/blob/d16af4a53aaa52210f8e444bb314a377aa0d868d/plugins/events-node/api-report.md?plain=1#L17) as well as a type, where all of the component parts probably are important in a lot of cases.

I think whatever interface we place here needs to receive raw `EventParams`, not an unpacked or simplified version of it. And it might be useful to keep the plain `supportsEventTopics(): string[]` method as-is in this interface too, and passing that through.

The `deltaMapper` method could just as well be named `onEvent` too at that point, but of course it needs to be able to return or call out to something that's specific to the incremental model.

We could return just a plain single delta as you suggest here, but I wonder if that's enough. Imagine that you listen for repo deletion events. At that point you probably need to issue some form of large sweeping deletion, and I'm not sure how to express something like that.

One possibility:

```ts
onEvent(params: EventParams): Promise<IncrementalChange> {
  return { added: ... };
}
```

Another:

```ts
onEvent(params: EventParams, ctx: SomeEventContextThing): Promise<void> {
  await ctx.delta({ added: ... });
}
```

In the latter we could add more types of "reaction" based on the event in future iterations. Not sure if premature optimization, but I was thinking that return types might be harder to evolve.

Thoughts?

Reply:
I like this. Again, my only real concern is that it might be out of scope for an entity provider. But maybe not. It's almost at that point turning into an entity _manager_. But that's not really a bad thing.

Reply:
This API makes a lot of sense.

Reply:
`onEvent` is actually already there! It's just in the engine, rather than in the entity provider.

Reply:
Sure. But it's good name for the incremental provider method too.

Reply:
Yeah, it's a good method name because it makes the purpose very clean and it's recognizable. I find `deltaMapper` kinda confusing.

Reply:
Done!